_posts/2019-12-08-what-is-a-shell.md

```diff
@@ -0,0 +1,166 @@
+
+Some examples of well known shells are Bash, Zsh, Fish, Powershell, Korn Shell, and csh.
+There are many others (I'm sorry if I didn't mention your favorite one!), but listing them all would belabor the point.
+
+### Example Usage
+
+In the [Bash shell](https://en.wikipedia.org/wiki/Bash_%28Unix_shell%29) (bourne again shell), 
+you can list directory contents using the command `ls`,
+where it displays information about the current directory by default.
+[Powershell](https://docs.microsoft.com/en-us/powershell/) on Windows provides users a similar functionality using the same `ls` command.
+
+In order to alter the way information is displayed,
+you may have to provide additional input to the command,
+which is called a **flag**.
+For example, `ls -l` provides a "long list" format which also shows file permissions, file creation and file access information.
+That particular flag will work in Bash and most other Bash inspired shells, but it doesn't work in Powershell.
+
+Powershell provides a similar features in the form of a cmdlet, pronounced _command-let_.
+You can try `ls` in Powershell and it will work, but also try `man ls` to get the manual page.
+You'll find that the cmdlet is actually named `Get-ChildItem`.
+If you use `alias ls` in Powershell,
+you will see the command is an alias for the Get-ChildItem.
+
+It's somewhat interesting that `ls` in Bash is similar to `ls -Name` in Powershell
+and `ls -l` in Bash is similar to the default `ls` in Powershell.
+
+There are plenty of other executable programs that operating systems have to offer.
+For Linux/MacOS,
+you most likely have the path `/usr/bin` in your `PATH` variable,
+which is what directories your shell will look in to find executable programs.
+Try `ls -l /usr/bin` to see a list of executables which you can run from the command line.
+If you're curious about any of them,
+all you need is the manual page by typing the executable name preceded by `man`.
+For example `man zip`.
+Similar functionality can be achieved in Powershell by using `Get-Command`.
+
+### Manipulating Data
+
+In the previous section we saw how a user may list the available commands or executables they can run via their shell.
+If you've been trying these on your own system,
+you may have noticed an overwhelming amount of output.
+Sometimes you may need all of it,
+but often you might want a subset of the data.
+
+We'll continue with `ls`,
+as it is fairly simple to work with.
+In Bash, if you want to `ls` only directories,
+or in other words, _filter entries by the type directory_,
+then you may use the command `ls -d */`.
+The logic behind this is that the `-d` flag will output the / character after a directory and then we're using a glob to match anything that ends with the /.
```

Comment:
That's not quite what's happening here. The glob gets evaluated first (try `echo */`). That gets expanded to a set of directories that get passed to `ls -d`. `-d` prevents it from descending into those directories.